guides/keras_cv/simsiam-with-kerascv.py

```diff
@@ -0,0 +1,604 @@
+representations is easier and requires fewer labels because the pre-trained model
+already produces meaningful and generally useful features.
+
+Overall, self-supervised pre-training learns representations which are more generic and
+robust than other approaches to augmented training and pre-training.  An overview of
+the general contrastive learning process is shown below:
+
+![Contrastive overview](https://i.imgur.com/mzaEq3C.png)
+
+In this tutorial, we will use the [SimSiam](https://arxiv.org/abs/2011.10566) algorithm
+for contrastive learning.  As of 2022, SimSiam is the state of the art algorithm for
+contrastive learning; allowing for unprecedented scores on CIFAR-100 and other datasets.
+
+To get started, we will sort out some imports.
+"""
+import resource
+
+low, high = resource.getrlimit(resource.RLIMIT_NOFILE)
+resource.setrlimit(resource.RLIMIT_NOFILE, (high, high))
+
+
+import gc
+import os
+import random
+import time
+import tensorflow_addons as tfa
+import keras_cv
+from pathlib import Path
+from tensorflow_similarity.layers import GeneralizedMeanPooling2D, MetricEmbedding
+import matplotlib.pyplot as plt
+import numpy as np
+from tensorflow.keras import layers
+from tabulate import tabulate
+import tensorflow_similarity as tfsim  # main package
+import tensorflow as tf
+from keras_cv import layers as cv_layers
+
+import tensorflow_datasets as tfds
+
+tfsim.utils.tf_cap_memory()  # Avoid GPU memory blow up
+tfds.disable_progress_bar()
+
+BATCH_SIZE = 512
+PRE_TRAIN_EPOCHS = 50
+VAL_STEPS_PER_EPOCH = 20
+WEIGHT_DECAY = 5e-4
+INIT_LR = 3e-2 * int(BATCH_SIZE / 256)
+WARMUP_LR = 0.0
+WARMUP_STEPS = 0
+DIM = 2048
+
+"""
+## Data Loading
+
+Next, we will load the STL-10 dataset.  STL-10 is a dataset consisting of 100k unlabelled
+images, 5k labelled training images, and 10k labelled test images.  Due to this distribution,
+STL-10 is commonly used as a benchmark for contrastive learning models.
+
+First lets load our unlabelled data
+"""
+train_ds = tfds.load("stl10", split="unlabelled")
+train_ds = train_ds.map(
+    lambda entry: entry["image"], num_parallel_calls=tf.data.AUTOTUNE
+)
+train_ds = train_ds.map(
+    lambda image: tf.cast(image, tf.float32), num_parallel_calls=tf.data.AUTOTUNE
+)
+train_ds = train_ds.shuffle(buffer_size=8 * BATCH_SIZE, reshuffle_each_iteration=True)
+
+"""
+Next, we need to prepare some labelled samples.
+This is done so that TensorFlow similarity can probe the learned embedding to ensure
+that the model is learning appropriately.
+"""
+
+(x_raw_train, y_raw_train), ds_info = tfds.load(
+    "stl10", split="train", as_supervised=True, batch_size=-1, with_info=True
+)
+x_raw_train, y_raw_train = tf.cast(x_raw_train, tf.float32), tf.cast(
+    y_raw_train, tf.float32
+)
+x_test, y_test = tfds.load(
+    "stl10",
+    split="test",
+    as_supervised=True,
+    batch_size=-1,
+)
+x_test, y_test = tf.cast(x_test, tf.float32), tf.cast(y_test, tf.float32)
```

Comment:
If I remember correctly, STL-10 has also got a split called "unlabeled" ([reference](https://www.tensorflow.org/datasets/catalog/stl10)). Would be cool to incorporate that for SSL pre-training. If not possible here in the example, maybe add a note?

Reply:
check out line 88!